celery_unique/core.py

```diff
@@ -0,0 +1,199 @@
+# -*- coding: utf-8 -*-
+from __future__ import absolute_import
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import datetime
+
+UNIQUE_KEY_PREFIX = 'celery_unique'
+
+
+class UniqueTaskMixin(object):
+    abstract = True
+    unique_key = None
+    unique_backend = None
+
+    def apply_async(self, args=(), kwargs={}, task_id=None, **options):
+        """
+        Apply tasks asynchronously by sending a message.
+
+        This method serves either as a wrapper for `celery.Task.apply_async()`
+        or, if the task decorator was configured with a `unique_backend`,
+        callable `unique_key` and `apply_async()` was called with either an
+        `eta` or `countdown` argument, the task will be treated as unique.
+
+        In these cases, this method will first revoke any extant task which
+        matches the same unique key configuration before proceeding to publish
+        the task.  Before returning, a unique task's identifying unique key
+        will be saved to Redis as a key, with its task id (provided by the
+        newly-created `AsyncResult` instance) serving as the value.
+
+        See ``celery.Task.apply_async()``
+
+        :param func unique_key: Function used to generate a unique key to
```

Comment:
Two things:
1. Instead of `func`, this maybe would be better represented as `types.FunctionType`
2. I'm not familiar with the Sphinx syntax of `:param <python type> <param name>:`; usually I see types documented separately with `:type <param name>: <python type>`.  This could very well be a valid syntax I'm just not used to, but at any rate, I think we should stay consistent with our usual style (bears mentioning that the docstrings in the `celery_unique.backends` module use `:param` and `:type`.